tomviz/Pipeline.cxx

```diff
@@ -110,25 +113,70 @@ void Pipeline::execute()
   execute(m_data);
 }
 
-void Pipeline::execute(DataSource* dataSource, Operator* start)
+void Pipeline::startedEditingOp(Operator* op)
+{
+  ++m_editingOperators;
+  op->setEditing();
+}
+
+void Pipeline::finishedEditingOp(Operator* op)
+{
+  if (op == nullptr) {
+    return;
+  }
+
+  if (op->isModified()) {
+    op->resetState();
+  } else {
+    op->setComplete();
+  }
+
+  if (m_editingOperators > 0) {
+    --m_editingOperators;
+    if (m_editingOperators == 0 && !isRunning()) {
+      execute(op->dataSource());
+    }
+  }
+}
+
+void Pipeline::execute(DataSource* ds, Operator* start)
 {
   if (paused()) {
     return;
   }
-  emit started();
 
-  if (dataSource == nullptr) {
-    dataSource = m_data;
+  if (ds == nullptr) {
+    ds = m_data;
   }
 
-  auto operators = dataSource->operators();
+  if (beingEdited(ds)) {
+    qDebug() << "Can't execute, edit in progress";
```

Comment:
If this is important for the user to see then it should appear in a dialog otherwise lets remove it.